datastore/longhorn.go

```diff
@@ -53,108 +53,138 @@ func (s *DataStore) UpdateCustomizedSettings(overrideDefaultSettings map[types.S
+	err := s.ValidateSetting(string(name), value)
+	if err != nil {
+		return err
+	}
+
 	setting, err := s.GetSettingExact(name)
 	if err != nil {
 		if !ErrorIsNotFound(err) {
 			return err
 		}
+
 		setting = &longhorn.Setting{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:        string(name),
-				Annotations: map[string]string{types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey): definition.ConfigMapResourceVersion},
+				Annotations: map[string]string{types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey): defaultSettingCMResourceVersion},
 			},
-			Value: definition.Default,
+			Value: value,
 		}
 
-		if err := s.ValidateSetting(string(setting.Name), setting.Value); err != nil {
+		if _, err = s.CreateSetting(setting); err != nil && !apierrors.IsAlreadyExists(err) {
 			return err
 		}
+		return nil
+	}
 
-		_, err := s.CreateSetting(setting)
-		if err != nil && !apierrors.IsAlreadyExists(err) {
-			return err
+	if setting.Annotations == nil {
+		setting.Annotations = map[string]string{}
+	}
+
+	setting.Annotations[types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey)] = defaultSettingCMResourceVersion
+	setting.Value = value
+	_, err = s.UpdateSetting(setting)
+	return err
+}
+
+func (s *DataStore) updateSettingDefinitionsWithCustomizedDefaultSettings(customizedDefaultSettings map[string]string, defaultSettingCM *v1.ConfigMap) error {
+	for _, sName := range types.SettingNameList {
+		definition, ok := types.GetSettingDefinition(sName)
+		if !ok {
+			return fmt.Errorf("BUG: setting %v is not defined", sName)
 		}
 
-		return nil
+		if value, ok := customizedDefaultSettings[string(sName)]; ok {
+			definition.Default = value
+			types.SetSettingDefinition(sName, definition)
+		}
 	}
+	return nil
+}
 
-	if setting.Value != definition.Default {
-		setting.Value = definition.Default
-		if setting.Annotations == nil {
-			setting.Annotations = map[string]string{}
+func (s *DataStore) syncDefaultSettingsWithSettingCRs(customizedDefaultSettings map[string]string, defaultSettingCMResourceVersion string) error {
+	for _, sName := range types.SettingNameList {
+		configMapResourceVersion := ""
+		if s, err := s.GetSettingExact(sName); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+		} else {
+			if s.Annotations != nil {
+				configMapResourceVersion = s.Annotations[types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey)]
+			}
 		}
-		setting.Annotations[types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey)] = definition.ConfigMapResourceVersion
 
-		if err := s.ValidateSetting(string(setting.Name), setting.Value); err != nil {
-			return err
+		definition, ok := types.GetSettingDefinition(sName)
+		if !ok {
+			return fmt.Errorf("BUG: setting %v is not defined", sName)
 		}
 
-		_, err = s.UpdateSetting(setting)
-		return err
+		value, exist := customizedDefaultSettings[string(sName)]
+		if !exist {
+			continue
+		}
```

Comment:
I don't understand why we still need the input `customizedDefaultSettings` or `definition, ok := types.GetSettingDefinition(sName)` here.

Reply:
Sorry. We need `customizedDefaultSettings`, because we only want to update the modified fields.

The purpose of `definition, ok := types.GetSettingDefinition(sName)` is only to know if the field is `Required` .